### PR TITLE
tradingview: init at 1.0.10

### PR DESCRIPTION
**Pull request by k3a**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/office/tradingview/default.nix [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    runHook preUnpack
+    unsquashfs "$src" "resources/app.asar" "resources/app.asar.unpacked" "meta/gui"
+    cd squashfs-root
+    runHook postUnpack
+  '';
+
+  installPhase =
+    ''
+      runHook preInstall
+
```
Comment (SuperSandro2000) [inline pkgs/applications/office/tradingview/default.nix:38]:
```suggestion
  installPhase = ''
    runHook preInstall

```
please adjust formatting

File: pkgs/applications/office/tradingview/default.nix [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  src = fetchurl {
+    url = "https://api.snapcraft.io/api/v1/snaps/download/nJdITJ6ZJxdvfu8Ch7n5kH5P99ClzBYV_${revision}.snap";
+    hash = "sha512-UZ8ogKdlmYkeIYC+skqxRPDd1KZYmPv+pbCW7XvvHsCs+ptkbOxaP+M82uHugmPf4aTUVU+0cI3hvjV2kTYkjA==";
+  };
+
+  nativeBuildInputs = [ squashfsTools makeWrapper ];
+
+  dontConfigure = true;
+  dontBuild = true;
+
```
Comment (SuperSandro2000) [inline pkgs/applications/office/tradingview/default.nix:24]:
```suggestion
```

File: pkgs/applications/office/tradingview/default.nix [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+, squashfsTools
+, makeWrapper
+, electron
+, libsecret
+}:
+
+stdenv.mkDerivation rec {
+  pname = "tradingview";
+  version = "1.0.13";
+  revision = "24";
```
Comment (SuperSandro2000) [inline pkgs/applications/office/tradingview/default.nix:13]: This should be in a let in, probably together with version

File: pkgs/applications/office/tradingview/default.nix [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      cp -R resources "$out/share/${pname}"
+      install -Dm644 meta/gui/${pname}.desktop "$out/share/applications/${pname}.desktop"
+      substituteInPlace $out/share/applications/${pname}.desktop \
+        --replace 'Icon=''${SNAP}/meta/gui/icon.png' 'Icon=${pname}' \
+        --replace 'Categories=Finance;' 'Categories=Office;Finance'
+      install -Dm644 meta/gui/icon.png $out/share/pixmaps/${pname}.png
+
+      makeWrapper ${electron}/bin/electron \
+        $out/bin/${pname} \
+        --add-flags $out/share/${pname}/resources/app.asar \
```
Comment (SuperSandro2000) [inline pkgs/applications/office/tradingview/default.nix:50]:
```suggestion
      makeWrapper ${electron}/bin/electron $out/bin/tradingview \
        --add-flags $out/share/tradingview/resources/app.asar \
```

File: pkgs/applications/office/tradingview/default.nix [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      runHook preInstall
+
+      # desktop file and icon
+      mkdir -p "$out/share/${pname}"
+      cp -R resources "$out/share/${pname}"
+      install -Dm644 meta/gui/${pname}.desktop "$out/share/applications/${pname}.desktop"
+      substituteInPlace $out/share/applications/${pname}.desktop \
+        --replace 'Icon=''${SNAP}/meta/gui/icon.png' 'Icon=${pname}' \
+        --replace 'Categories=Finance;' 'Categories=Office;Finance'
+      install -Dm644 meta/gui/icon.png $out/share/pixmaps/${pname}.png
```
Comment (SuperSandro2000) [inline pkgs/applications/office/tradingview/default.nix:46]:
```suggestion
      mkdir -p "$out/share/tradingview"
      cp -R resources "$out/share/tradingview"
      install -Dm644 meta/gui/tradingview.desktop "$out/share/applications/tradingview.desktop"
      substituteInPlace $out/share/applications/tradingview.desktop \
        --replace 'Icon=''${SNAP}/meta/gui/icon.png' 'Icon=tradingview' \
        --replace 'Categories=Finance;' 'Categories=Office;Finance'
      install -Dm644 meta/gui/icon.png $out/share/pixmaps/tradingview.png
```

File: pkgs/applications/office/tradingview/update.sh [hunk 6]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i bash -p curl gnused gnugrep jq
+
+set -eu -o pipefail
+
+dirname=$(dirname "$0" | xargs realpath)
```
Comment (SuperSandro2000) [inline pkgs/applications/office/tradingview/update.sh:6]:
```suggestion
dirname="$(realpath "$(dirname "$0")")"
```

File: pkgs/applications/office/tradingview/update.sh [hunk 7]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i bash -p curl gnused gnugrep jq
+
+set -eu -o pipefail
```
Comment (SuperSandro2000) [inline pkgs/applications/office/tradingview/update.sh:4]:
```suggestion
set -eou pipefail
```

File: pkgs/applications/office/tradingview/update.sh [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+channel="${1:-stable}" # stable/candidate/edge
+
+# create bash array from snap info
+snap_info=($(
+  curl -s -H 'X-Ubuntu-Series: 16' \
+    "https://api.snapcraft.io/api/v1/snaps/details/tradingview?channel=$channel" \
+  | jq --raw-output \
+    '.revision,.download_sha512,.version,.last_updated'
+))
```
Comment (SuperSandro2000) [inline pkgs/applications/office/tradingview/update.sh:19]:
```suggestion
snap_info=($(
  curl -s -H 'X-Ubuntu-Series: 16' "https://api.snapcraft.io/api/v1/snaps/details/tradingview?channel=$channel" \
  | jq --raw-output '.revision,.download_sha512,.version,.last_updated'
))
```

File: pkgs/applications/office/tradingview/update.sh [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+# "revision" is the actual version identifier on snapcraft, the "version" is
+# just for human consumption. Revision is just an integer that gets increased
+# by one every (stable or unstable) release.
+upstream_revision="${snap_info[0]}"
+sha512="${snap_info[1]}"
+upstream_version="${snap_info[2]}"
+last_updated="${snap_info[3]}"
+sri_hash=$(nix hash to-sri --type sha512 $sha512)
```
Comment (SuperSandro2000) [inline pkgs/applications/office/tradingview/update.sh:28]: Why are we using sha512? Standard is sha256.

Also why are we converting it in the first place? If upstream has a sha512 that works then we should use it without base64 encoding. This makes things a lot easier to double check.

**Comment by k3a** [comment]
Unfortunately I have to close this PR because TradingView did something in the version 1.0.17 (revision 28) which makes it incompatible with the stock Electron. 

It displays `(Error):Local service with contracts [tvd:tab-drag-handler] not registered, and remoting not configured. Instance can not be created.` message on strartup with an `Ok` button to close the app.

But there is a nice workaround - probably better than their official "desktop app" - just use a web shortcut in a Chromium:
1. Open https://www.tradingview.com/chart in Chromium.
2. Open threedot menu, select More Tools -> Create Shortcut... select a separate window and create the shortcut.
3. Then, because of their long-standing [chrome bug](https://bugs.chromium.org/p/chromium/issues/detail?id=118613), one needs to edit the shortcut Exec line by adding `--user-data-dir` and URL again like this:
`Exec=chromium --user-data-dir=/home/username/.config/google-chrome/TradingView --class=crx_fbamlndehdinmdbhpcihcihhmjmmpgjn https://www.tradingview.com/chart/` 
4. That will make a working shortcut for opening a separate browser window with its own icon in the taskbar, just like their original 'app'. And this one will be compatible forever.

**Pull request closed without merging**